lib/dao.js

```diff
@@ -624,7 +624,184 @@ DataAccessObject.upsert = function(data, options, cb) {
   }
   return cb.promise;
 };
+/**
+ * Performs upsert based on the instance that matches the where clause
+ . This is a PUT operation and based on
+ * non-primary key. The filter takes one key value pair and updates single matching instance.*
+ */
+DataAccessObject.patchOrCreateWithWhere =
+  DataAccessObject.upsertWithWhere = function(where, data, options, cb) {
+    var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+    if (connectionPromise) { return connectionPromise; }
+    assert(arguments.length >= 1, 'At least one argument is required');
+    if (data === undefined && options === undefined && cb === undefined && arguments.length === 1) {
+      data = where;
+      where = {};
+    } else if (options === undefined && cb === undefined) {
+      if (typeof data === 'function') {
+        cb = data;
+        data = where;
+        where = {};
+      }
+    } else if (cb === undefined) {
+      if (typeof options === 'function') {
+        cb = options;
+        options = {};
+      }
+    }
+    data = data || {};
+    options = options || {};
+    cb = cb || utils.createPromiseCallback();
+    assert(typeof where === 'object', 'The where argument must be an object');
+    assert(typeof data === 'object', 'The data argument must be an object');
+    assert(typeof options === 'object', 'The options argument must be an object');
+    assert(typeof cb === 'function', 'The cb argument must be a function');
+    var hookState = {};
+    var self = this;
+    var Model = this;
+    var connector = Model.getConnector();
+    var modelObject = connector._models;
+    var modelName = Object.keys(modelObject)[0];
```

Comment:
`modelRegistry` would be a better variable name.

However, I am afraid these two lines are based on a wrong assumption. You should get the model name this way:

```js
var modelName = Model.modelName;
```